File: 91-decode-ways/decode-ways.java
class Solution {
    public int numDecodings(String s) {
        int prev = 1;
        int prev2 = 0;
        int curr = 0;

        for (int i = s.length() - 1; i >= 0; i--) {
            if (s.charAt(i) == '0') {
                curr = 0;

            } else {
                int step1 = prev;
                int step2 = 0;
                if (i + 1 < s.length() && (s.charAt(i) == '1' || (s.charAt(i) == '2' && s.charAt(i + 1) < '7'))) {
                    step2 = prev2;
                }

                curr = step1 + step2;

            }
            prev2 = prev;
            prev = curr;

        }

        return curr;

    }
}

File: 91-decode-ways/decode-ways.py
class Solution:
    def numDecodings(self, s: str) -> int:
        n = len(s)
        dp = [-1]*n 

        def dfs(i):
            if i >= n:
                return 1

            if s[i] == "0":
                return 0
            
            if dp[i] != -1:
                return dp[i]

            take = dfs(i + 1)
            not_take = 0

            if (
                s[i] != "0"
                and i + 1 < n
                and ((s[i] == "2" and s[i + 1] <= "6") or s[i] < "2")
            ):
                not_take = dfs(i + 2)

            dp[i] = take + not_take
            return dp[i]

        return dfs(0)
